File: 13.Dynamic_Programming/01.coins.cpp
# B站 https://www.bilibili.com/video/av927393080/


//dp 

using namespace std;
class Solution{
public:
    int coinChange(vector<int>&coins,int amount){
        //初始化数组dp,大小为 amount+1； # indx 作为求解数值，内容为最优解结果
        vector<int>dp(amount + 1 , -1);//全部数值初始化为-1，作为没有计算最优解的标记
        dp[0]=0  //初始化，0的最优解就是0
        //计算金额1至amount的最优解
        for(int i = 1;i<=amount;i++){ 
            //根据硬币类型总数，0开始index
            for(int j=0;j<coins.size();j++){
                // coins[j] 金额 小于 目标数值i ,并且，dp最优解中 1-当前金额的数值 在dp 中有存储最优解答案
                if(coins[j]<=i && dp[i-coins[j]]!=-1){
                    
                    if(dp[i]==-1|| dp[i]>dp[i-coins[j]]+1){
                        dp[i]=dp[i-coins[j]]+1;//更新dp[i]
                }
                }
            }
        }
        return dp[amount];
    }
};


int main(){
    vector<int>coins;//设置面值大小为(1,2,5,7,10)
    coins.push_back(1)
    coins.push_back(2)
    coins.push_back(5)
    coins.push_back(7)
    coins.push_back(10)
    Solution solution;
    printf("%d\n",solution.coinChange(coins,14))
    return 0;
}

File: 13.Dynamic_Programming/01.coins.py
# B站 https://www.bilibili.com/video/av927393080/
# 根据理解修改python版

#动态规划的思想，就是最优解 拆分为最有1步骤 + 前面n-1 步骤的最优解
#从最小的开始开始推演

amount=63 # 目标金额
coins=[1,2,5,7,10] #硬币金额
dp = [-1]*(amount+1) #建立0-amount位置长度的 数组向量
dp[0] = 0 # 金额0的最优解是0个coins
# 金额为1 的最优解  ，拆分为 1金额+ 0金额的最优解dp[0]
# 从[1,2,5,7,10] 中任意一个硬币 + dp[0]
# 只有硬币 1金额的硬币1 可以， 需要的硬币就是1+dp[] = 1 + 0 =1

# 金额为3的 最优解，可用硬币小于等于3 ，【1，2】可以拆分为  硬币1 + dp[2] 或者 硬币2 + dp[1] 可以得到3
# 可以解 为  【硬币1或者硬币2】 + dp[2/1] = 1 + 1 = 2

# 硬币问题,从0开始，迭代，找最优解，最
for i in range(amount+1): # 目标的最优解，拆分为0-amount 的每一个的最优解
    #
    for j in range(len(coins)):#循环所有硬币，指定1 个硬币金额 coins[j]，作为解 dp[i-coins[j]]+1
        #硬币金额效于目标，并且，减去硬币金额后的 金额数在 dp 数组向量中有最优解存储
        if coins[j]<=i and dp[i-coins[j]]!=-1:
            
            #假设 dp[i-coins[j]] + 1枚硬币
            if dp[i]==-1 or dp[i]>dp[i-coins[j]]+1:
                        dp[i]=dp[i-coins[j]]+1
    print(dp)
    print(len(dp))
    print(dp[amount])
